Xqtl refactor

cynthiaewu opened the pull request:
xQTL:

 For pvalue calculation:
- Added eigendecomposition method for calculating empirical null that utilizes existing multithreading for cpma/xqtl calculations. Writes the null values to files. These files can be utilized again if user wants to reuse the simulated null values for another xQTL run (not implemented)
- Added chi2 method

Simulate:

- Added an instance check for cov to prevent warning messages with --gxg-corr option

*Changed all tabs to spaces

gymreklab commented:
Notes:

* `zscores = data.pivot_table(index='SNP', columns='gene', values='t-stat')`: this will be slow or not possible for very large datasets. We should consider having a way to default to only reading up to a maximum number of genes or something.
* We should look into hdf5 format for storing these huge matrices in a more efficiently accessible way. 
* I am confused by the multi-threading part for eigendecomposition. It looks like the processes use the same "worker" and "writer" as the main jobs. But "simulate_null_values" does some work already to get the simulated z-scores before adding jobs to the queue. I am not sure how slow that is or if it is worth having those steps also computed by the workers.
* After computing the null values, we then re-read them from a file. Is this slow? Should we instead just keep them in memory?
* We should only read from the "null_sim.tab" files (line 220) if args.null_method is "eigen"

cynthiaewu commented:
If --null_method eigen:
Added a check to see if user's empirical null values file exists. If it does, read the values directly from file. If not, simulate null values through eigendecomposition, keep the values in memory to use for next step, and write the values to file for future usage. 

multithreading for the simulate_null_values is still useful. It does some work already to get the simulated z-scores in chunks (currently set to 30,000) which is reasonably fast. However, having to compute CPMA and xQTL values for these simulated z-scores are slow. We need to calculate 500,000 cpma/xQTL values which is probably much more than the cpma/xQTL values we will need to calculate for the user's input snps. Since we're using the multithreading part for the smaller number of user's input snps, it seems to make sense to also use the multithreading for the larger number of simulated values.

> We should only read from the "null_sim.tab" files (line 220) if args.null_method is "eigen"

Fixed

> We should look into hdf5 format for storing these huge matrices in a more efficiently accessible way.

We don't currently save any huge matrices, but can look into this. We save the null values (500,000 values each for CPMA and xQTL) in a file, and the CPMA/xQTL results for the user's input file.

> zscores = data.pivot_table(index='SNP', columns='gene', values='t-stat'): this will be slow or not possible for very large datasets. We should consider having a way to default to only reading up to a maximum number of genes or something.

Yes, this will be needed. We should think of a way to grab lines randomly and what the maximum number of genes should be.

gymreklab commented:
Thanks for these changes. Merging!
